### ISSUE TITLE
conda package

### ISSUE THREAD

**Issue by gdementen** [Roche/pyreadstat#16]
If you could create a conda package that would make my life easier (so far so that I can continue to update my "bleeding edge" environment with "update --all").

BTW: Thanks for the very nice library, even though on the file (~130Mb) I tested it today it wasn't any faster than pandas 0.24 (both were fast), the builtin encoding autodetection saved me quite a bit of trial and errors.

**Comment by ofajardo** [comment]
Regarding conda, I have pip with is more general (works both for virtualenv and conda). What about something like [this]
(https://stackoverflow.com/questions/33187676/update-all-pip-packages-that-dont-come-from-conda)
```
conda list | grep "<pip>" | cut -d " " -f 1 | xargs pip install --upgrade
```
Would that work for you?

Regarding speed, what kind of file are you reading? I only benchmarked SAS. For SPSS and STATA I did not do any benchmarks.

**Comment by gdementen** [comment]
Thanks a lot for your effort in trying to find a solution, but I am on
Windows and even if I have some ported gnu tools installed (I missed grep
too much from my linux days), I don't have a full blown unix shell at my
disposal. I could surely find a way to do the same on Windows but it's just
too much trouble. Now I don't expect anything from you (!) but for what
it's worth, nowadays there are more and more people who do not even
consider using libraries without conda packages.

Regarding the speed, that's a sas7bdat file with not that many rows but
many columns (out of my head something like 17000 rows but 1020 columns),
so that's probably not a typical row/columns ratio.

**Comment by ofajardo** [comment]
All right, I'll take a look to creating a conda package. I am waiting for some changes in ReadStat to do a new release, I check once that's ready. 
I am very happy about you using the package and contributing to it. I am just curious: if speed is not the advantage for you vs pandas, what is then the key features that make you prefer this package over pandas? Thanks for the feedback.

**Comment by gdementen** [comment]
I don't want to pressure you into doing anything, I am just saying it would be appreciated.

We don't use pyreadstat in our production workflow yet but I am testing it. For the story, I installed it initially for the speed promise because several colleagues of mine receive some big databases from external sources as sas7bdat files but want to work on them outside of SAS and it currently takes quite a bit of time to convert them to other formats (HDF5 or SQL) using pandas. But I have been on urgent projects since then and haven't found the time to actually try pyreadstat on our big files yet. Then this morning a colleague had an issue with the sas7bdat file I talk above which failed to load using pandas so I tried loading it with pyreadstat and it worked. But then I noticed upgrading pandas to 0.24 worked too and was a tad faster so I thought I would mention that...

**Comment by gdementen** [comment]
Yes, but only for pure-python packages and that was relatively easy. I
suppose packages with C/Cython extensions are a bit trickier.

Now I think that since you already have a setup.py, you should be able to
piggy back on it.

See https://conda.io/projects/conda-build/en/latest/source/recipe.html

"conda skeleton pyreadstat" should help getting you started.

I hope it helps...

**Comment by ofajardo** [comment]
Right ... so that was what I was looking at and basically as you said compiling from source in this case would be very very tricky. However, I found that you can build the conda package from a python wheel. 

I am trying to get it into [conda-forge](https://github.com/conda-forge/staged-recipes/pull/7675) but for now it is failing, does not want to install from pip. In the documentation they mention that packages should compile from source, so maybe it won't work. The nice thing about conda-forge is that they build the conda package for windows, mac and linux.

If this does not work, I think that the only thing I could do is to prepare the recipe (using pip), and put it in here, and then people would have to download the recipe, build the package in their machines and then finally do conda install. Does it sound OK?

alternatives?

**Comment by gdementen** [comment]
Honestly, in my experience it was much harder to get my packages working (and keep them so!) on conda-forge than to have them built locally and upload them myself to my own channel on anaconda.org.

So you might want to try that instead, even if that is not as slick. If that does not work, try uploading the wheels converted to conda packages to your own channel. That would hopefully work.

Avoiding to have to build "complicated packages" from source is one of the reasons conda is popular, so having the recipe but needing to build it ourselves seems to be a step backwards compared to the current wheel + using pip inside a conda environment solution. I guess that if the above alternatives do not work for you, you should just close this issue and forget about it unless somebody with the right expertise steps up to do it.

**Comment by ofajardo** [comment]
_If that does not work, try uploading the wheels converted to conda packages to your own channel. That would hopefully work._

Yes. But I can do only for windows and people will start asking for mac and linux which I cannot do, so they would have to build the conda package themselves. And because it is some extra work to upload the windows packages to the channel (because I would need to do for several python versions by hand) I would prefer that everyone does it by him or herself.

_Avoiding to have to build "complicated packages" from source is one of the reasons conda is popular, so having the recipe but needing to build it ourselves seems to be a step backwards compared to the current wheel + using pip inside a conda environment solution_

You will not have to compile from source, the recipe will just fetch the pre-compiled wheel from pypi and transform that into a conda package so that you can do conda update all. That is the only thing it would solve, for people who think it is an advantage (as you do).

Would it help?

_in my experience it was much harder to get my packages working (and keep them so!) on conda-forge_ 

Yeah, looks difficult. I manage to get the conda packages to build, but now tests fail. I have filed an [issue](https://github.com/conda-forge/staged-recipes/issues/7682) to see if somebody helps, let's see how that goes ... but indeed looks cumbersome.

**Comment by ofajardo** [comment]
Ok now I realize that what I am proposing regarding the recipe is not that great: people will have to build themselves every time there us a new release on pypi, so conda update all won't really work automatically. It still helps if there is an absolute neef to have it as conda package.

If that does not help, and conda forge either then I think no conda packages for now. 

But let's see conda forge ... there is some hope there.

**Comment by ofajardo** [comment]
Good news! After lot of trial and error and with the kind help of the conda forge people I managed to get the package to be built on all platforms. The recipe has been then admitted. In few hours to few days it should be available in conda forge. 

I'll update once it's ready.

**Comment by ofajardo** [comment]
Package now available on conda-forge (version 0.2.1). I am waiting for some changes in Readstat before doing a new release, once ready I will update on conda and pypi. 

You can install like this:

```
conda install -c conda-forge pyreadstat 
```

Thanks for the suggestion.

**Comment by gdementen** [comment]
Thanks a lot for your work!
